app/services/KeycloakProvider.scala

```diff
@@ -0,0 +1,92 @@
+package services
+
+import play.api.libs.ws.WS
+import play.api.{Application, Logger}
+import play.api.libs.json.JsObject
+import securesocial.core._
+import scala.collection.JavaConverters._
+
+
+/**
+ * A Keycloak OAuth2 Provider
+ */
+class KeycloakProvider(application: Application) extends OAuth2Provider(application) {
+  val Error = "error"
+  val Message = "message"
+  val Type = "type"
+  val Sub = "sub"
+  val Name = "name"
+  val GivenName = "given_name"
+  val FamilyName = "family_name"
+  val Picture = "picture"
+  val Email = "email"
+  val Groups = "isMemberOf"
+
+
+  override def id = KeycloakProvider.Keycloak
+
+  def fillProfile(user: SocialUser): SocialUser = {
+    val UserInfoApi = loadProperty("userinfoUrl").getOrElse(throwMissingPropertiesException())
+    val accessToken = user.oAuth2Info.get.accessToken
+    val promise = WS.url(UserInfoApi.toString).withHeaders(("Authorization", "Bearer " + accessToken)).get()
+
+    try {
+      val response = awaitResult(promise)
+      val me = response.json
+      Logger.debug("Got back from Keycloak : " + me.toString())
+      (me \ Error).asOpt[JsObject] match {
+        case Some(error) =>
+          val message = (error \ Message).as[String]
+          val errorType = ( error \ Type).as[String]
+          Logger.error("[securesocial] error retrieving profile information from Keycloak. Error type = %s, message = %s"
+            .format(errorType,message))
+          throw new AuthenticationException()
+        case _ =>
+          val userId = (me \ Sub).as[String]
+          val firstName = (me \ GivenName).asOpt[String]
+          val lastName = (me \ FamilyName).asOpt[String]
+          val fullName = (me \ Name).asOpt[String]
+          val avatarUrl = ( me \ Picture).asOpt[String]
+          val email = ( me \ Email).asOpt[String]
+          val groups = ( me \ Groups).asOpt[List[String]]
```

Comment:
does the groups work, do we get this information? Do we want to use roles instead?

Reply:
do we want to just add roles as well, i think it can just be a copy and paste, and I think roles are already passed along.

Reply:
I don't think social user object allow roles
```
SocialUser(identityId: IdentityId, firstName: String, lastName: String, fullName: String, email: Option[String],
                      avatarUrl: Option[String], authMethod: AuthenticationMethod,
                      oAuth1Info: Option[OAuth1Info] = None,
                      oAuth2Info: Option[OAuth2Info] = None,
                      passwordInfo: Option[PasswordInfo] = None) extends Identity
```

Reply:
sorry meant, do you want to have the check to see if the user is part of a role as well as a group to see if they can login?